File: Shared/Errors/ALTWrappedError.m
//
//  ALTWrappedError.m
//  AltStoreCore
//
//  Created by Riley Testut on 11/28/22.
//  Copyright © 2022 Riley Testut. All rights reserved.
//

#import "ALTWrappedError.h"

@implementation ALTWrappedError

+ (BOOL)supportsSecureCoding
{
    // Required in order to serialize errors for legacy AltServer communication.
    return YES;
}

- (instancetype)initWithError:(NSError *)error userInfo:(NSDictionary<NSString *,id> *)userInfo
{
    self = [super initWithDomain:error.domain code:error.code userInfo:userInfo];
    if (self)
    {
        if ([error isKindOfClass:[ALTWrappedError class]])
        {
            _wrappedError = [(ALTWrappedError *)error wrappedError];
        }
        else
        {
            _wrappedError = [error copy];
        }
    }
    
    return self;
}

- (NSString *)localizedDescription
{
    NSString *localizedFailureReason = self.wrappedError.localizedFailureReason ?: self.wrappedError.localizedDescription;
    
    NSString *wrappedLocalizedDescription = self.wrappedError.userInfo[NSLocalizedDescriptionKey];
    if (wrappedLocalizedDescription != nil)
    {
        NSString *localizedFailure = self.wrappedError.userInfo[NSLocalizedFailureErrorKey];
        
        NSString *fallbackDescription = localizedFailure != nil ? [NSString stringWithFormat:@"%@ %@", localizedFailure, localizedFailureReason] : localizedFailureReason;
        if (![wrappedLocalizedDescription isEqualToString:fallbackDescription])
        {
            return wrappedLocalizedDescription;
        }
    }
    
    NSString *localizedFailure = self.userInfo[NSLocalizedFailureErrorKey];
    if (localizedFailure != nil)
    {
        NSString *localizedDescription = [NSString stringWithFormat:@"%@ %@", localizedFailure, localizedFailureReason];
        return localizedDescription;
    }
    
    // localizedFailure is nil, so return wrappedError's localizedDescription.
    return self.wrappedError.localizedDescription;
}

- (NSString *)localizedFailureReason
{
    return self.wrappedError.localizedFailureReason;
}

- (NSString *)localizedRecoverySuggestion
{
    return self.wrappedError.localizedRecoverySuggestion;
}

- (NSString *)debugDescription
{
    return self.wrappedError.debugDescription;
}

- (NSString *)helpAnchor
{
    return self.wrappedError.helpAnchor;
}

@end

File: Shared/Errors/ALTWrappedError.swift
//
//  ALTWrappedError.swift
//  AltStore
//
//  Created by Riley Testut on 10/18/22.
//  Copyright © 2022 Riley Testut. All rights reserved.
//

import Foundation

public class ALTWrappedError: NSError
{
    public let wrappedError: Error
    
    private var wrappedNSError: NSError {
        self.wrappedError as NSError
    }
    
    public init(error: Error, userInfo: [String: Any])
    {
        self.wrappedError = error
        
        super.init(domain: error._domain, code: error._code, userInfo: userInfo)
    }
    
    public required init?(coder: NSCoder)
    {
        fatalError("ALTWrappedError does not support NSCoding.")
    }
    
    override public var localizedDescription: String {
        if let localizedFailure = self.userInfo[NSLocalizedFailureErrorKey] as? String
        {
            let localizedFailureReason = self.wrappedNSError.localizedFailureReason ?? self.wrappedError.localizedDescription
            
            let localizedDescription = localizedFailure + " " + localizedFailureReason
            return localizedDescription
        }
        
        // localizedFailure is nil, so return wrappedError's localizedDescription.
        return self.wrappedError.localizedDescription
    }
    
    override public var localizedFailureReason: String? {
        return self.wrappedNSError.localizedFailureReason
    }
    
    override public var localizedRecoverySuggestion: String? {
        return self.wrappedNSError.localizedRecoverySuggestion
    }
    
    override public var debugDescription: String {
        return self.wrappedNSError.debugDescription
    }
    
    override public var helpAnchor: String? {
        return self.wrappedNSError.helpAnchor
    }
}
